SparseData

psomhorst opened the pull request:
This PR 
- fills out SparseData;
- creates IntervalData;
- adapts loading of Timpel data to store indices of detected min/max values and QRS complexes in SparseData and breaths as IntervalData;

This PR is based of off https://github.com/EIT-ALIVE/eitprocessing/pull/177 which fixed time slicing. 

Closes #167.

psomhorst commented:
@DaniBodor 
I think I'm almost done with SparseData and IntervalData, two types of dataset to hold non-continuous data. Before finalising, could you go through and comment on both classes? Is it clear what the goal of each class is? I think we should later add some proper documentation on what types of data to store in what class.

DaniBodor commented:
>   * fills out SparseData;
>   * creates IntervalData;

`SparseData`, as a name, is not very clear. How about `PunctualData` or `TimepointData` or something like that?

psomhorst commented:
> `SparseData`, as a name, is not very clear. 

Sparse data is an actual concept in data science. 
It refers to a dataset in which the variables with actual useful data are relatively rare. It is often contrasted with missing data, where data that should be known is unknown. See e.g. https://www.baeldung.com/cs/missing-vs-sparse-data. 

![quicklatex com-3424d2cea95a4ebc80b71cc8dc5532e6_l3](https://github.com/EIT-ALIVE/eitprocessing/assets/127968566/4b0e7094-eeb4-4075-93b3-54f3f4a1d798)

So for me, using the term sparse makes sense here. We don't represent unknown data as 0, but just omit them. Other than that, our implementation conforms to the description of sparse data.

psomhorst commented on eitprocessing/datahandling/breath.py:
```diff
@@ -0,0 +1,9 @@
+from typing import NamedTuple
+
+
+class Breath(NamedTuple):
```
I'm not sure whether that makes sense to me. Labelling all individual breaths is not very useful. It might be useful to label specific breaths at some point, e.g. to indicate which breath in sequence a corresponds to which breath in sequence b. 

I'm currently working on an algorithm using breaths, and I don't think numbering of sequential breaths is necessary. This also introduces extra complexity. E.g.: do you reset the numbering when slicing? Do you reset the numbering when concatenating? Etc. 

The algorithms that was used can be attached to the IntervalData object that contains these Breaths. The code I'm currently using in my project is below. 

```python
breath_detection_params = {
    "framerate": sequence.eit_data["raw"].framerate,
    "minimum_distance": 0.5 / respiratory_rate,
}
breath_detection = BreathDetection(**breath_detection_params)

breaths = breath_detection.find_breaths(
    peep_seq.continuous_data["global_impedance_(mdn)"].values
)
starts, _, ends = zip(*breaths, strict=True)
peep_seq.interval_data.add(
    IntervalData(
        "detected_breaths",
        "Detected breaths",
        None,
        "breaths",
        parameters={BreathDetection: breath_detection_params},
        time_ranges=list(
            zip(time[list(starts)], time[list(ends)], strict=True)
        ),
        values=breaths,
    )
)
```

psomhorst commented:
> We could of course redefine `SparseData` as having 0-value (or False) at all timepoints other timepoints, and I think then it does fit in the definition. It would be adding a lot of overhead though, which seems unnecessary for most purposes.

I would suggest NaN in that case. I can imagine data stored as SparseData that will have the value 0 (e.g. increase of end-expiratory volume during a breath). numpy handles NaN data much better than False data (False is converted to 0 when doing calculations, screwing up things like averages, while numpy has specific functions to handle NaN data, like `np.nanmax()`). 

I still think, though, that this data is by definition sparse data. The way we store it is slightly different than this article suggests. I don't think we should name it differently because it's stored slightly differently. It would be trivial to convert this data to one suitable for machine learning algorithms by adding a method 'to_ML_matrix()` or something like that.

As I see it there are two options:
- leave it as is (with this name)
- convert it to have a continuous time axis with mostly NaN values

This second option would make some things simpler and others more difficult. I don't think it's worth the extra work.

psomhorst commented:
@DaniBodor 

When working on this I figured out another issue we will run in sooner or later.

It does not make sense anymore to slice a sequence by index. The reason is that we let go of the idea that all datasets inside a sequence have to share a time axis. EIT data, continuous data, sparse data and interval data can all have different time axes. Intervaldata does not even have a real time axis, but has time ranges. When slicing a sequence by index, it is not clear what time that index refers to, so it is not possible to slice multiple datasets.

I think we can remove the SelectByIndex from Sequence. This will introduce some different challenges. E.g. an algorithm that selects a period by index will now have to find out what time points are associated with that index. I don't think this is a huge problem, though.

psomhorst commented on eitprocessing/datahandling/sparsedata.py:
```diff
@@ -1,2 +1,70 @@
+    Sparse data can consist of only time (e.g. detected QRS complexes) or time-value pairs (e.g. tidal impedance
+    variation at the end of each breath).
+
+    Values will generally be numeric values in arrays, but can also be lists of different types of object.
```
I updated the docstring in abee1fbbb780cff8c9297c1df24342d40681b136.

DaniBodor commented:
> > We could of course redefine `SparseData` as having 0-value (or False) at all timepoints other timepoints, and I think then it does fit in the definition. It would be adding a lot of overhead though, which seems unnecessary for most purposes.
> 
> I would suggest NaN in that case. I can imagine data stored as SparseData that will have the value 0 (e.g. increase of end-expiratory volume during a breath). I can also imagine SparseData having a boolean value, e.g. one indicating whether a breath was accidentally triggered by the ventilator itself.
> 
> numpy handles NaN data much better than False data (False is converted to 0 when doing calculations, screwing up things like averages, while numpy has specific functions to handle NaN data, like `np.nanmax()`).
> 
> I still think, though, that this data is by definition sparse data. The way we store it is slightly different than this article suggests. I don't think we should name it differently because it's stored slightly differently. It would be trivial to convert this data to one suitable for machine learning algorithms by adding a method 'to_ML_matrix()` or something like that.
> 
> As I see it there are two options:
> 
>   * leave it as is (with this name)
> 
>   * convert it to have a continuous time axis with mostly NaN values
> 
> 
> This second option would make some things simpler and others more difficult. I don't think it's worth the extra work.

OK, go for it.

DaniBodor commented:
> @DaniBodor
> 
> When working on this I figured out another issue we will run in sooner or later.
> 
> It does not make sense anymore to slice a sequence by index. The reason is that we let go of the idea that all datasets inside a sequence have to share a time axis. EIT data, continuous data, sparse data and interval data can all have different time axes. Intervaldata does not even have a real time axis, but has time ranges. When slicing a sequence by index, it is not clear what time that index refers to, so it is not possible to slice multiple datasets.
> 
> I think we can remove the SelectByIndex from Sequence. This will introduce some different challenges. E.g. an algorithm that selects a period by index will now have to find out what time points are associated with that index. I don't think this is a huge problem, though.

Indeed, without a dedicated time axis to `Sequence` it's odd to slice by index. 
That being said, it is a very useful operation that users will be used to using for other similar data. Hadn't we discussed at some point to tag one data source as `primary_data` or something, defaulting to `EITData` if present? If so, that could be the axis for slice by index. We can also allow an optional argument to the select by index method, indicating which data's time axis to use for it.

DaniBodor commented:
I will unsubscribe to this PR for now while it's still in draft stage. Please tag me or a request review when it's ready to make sure I get notified again.

psomhorst commented:
@DaniBodor This PR is now ready for review. 

I have finalised the functionality of SparseData and IntervalData. The tests from #175 work (except for one). The loading of medibus data still has to be adapted, but that will be addressed with #194 (which I already started work on).

DaniBodor reviewed (changes requested):
Lots of work done here! Great job.

I had largely finished reviewing this yesterday end of day. I now see that there was a rebase since then and didn't really compare what has changed since. I guess it's just a rebase because other PRs got merged. If there's some other change, please let me know and I can take another look at that part.

Really good PR. I added some comments to some of the parts. In addition, it would be good to add some additional tests to ensure that these data containers are functioning the way they are intended.
In the meantime, I will look more closely at our coverage implementation, which should be able to give us some pointers on which parts of the code are or aren't covered by the existing tests, which I think would be quite informative here.

Again, great job here already!

psomhorst commented on eitprocessing/datahandling/breath.py:
```diff
@@ -0,0 +1,9 @@
+from typing import NamedTuple
+
+
+class Breath(NamedTuple):
```
A `Breath` is not a subtype of `IntervalData`. `IntervalData` is a container that can contain objects like `Breath`.

psomhorst commented:
@DaniBodor Thanks for the great comments!

The rebase was indeed due to the merged changes with #189 and #200. 

Coverage was also next on my list of important tasks. I'll make a report and see what is not covered for IntervalData and SparseData.

psomhorst commented on eitprocessing/datahandling/datacollection.py:
```diff
@@ -5,14 +5,16 @@
 
-V = TypeVar("V", EITData, ContinuousData, SparseData)
+V_classes = (EITData, ContinuousData, SparseData, IntervalData)
+V = TypeVar("V", EITData, ContinuousData, SparseData, IntervalData)
```
This is what I did at first, but got an error. But now it works... Weird ;)

psomhorst commented on eitprocessing/datahandling/eitdata.py:
```diff
@@ -111,43 +105,14 @@ def _sliced_copy(
     def __len__(self):
         return self.pixel_impedance.shape[0]
 
-    @property
-    def global_baseline(self) -> np.ndarray:
-        """Return the global baseline, i.e. the minimum pixel impedance across all pixels."""
-        return np.nanmin(self.pixel_impedance)
-
-    @property
-    def pixel_impedance_global_offset(self) -> np.ndarray:
-        """Return the pixel impedance with the global baseline removed.
-
-        In the resulting array the minimum impedance across all pixels is set to 0.
-        """
-        return self.pixel_impedance - self.global_baseline
-
-    @property
-    def pixel_baseline(self) -> np.ndarray:
-        """Return the lowest value in each individual pixel over time."""
-        return np.nanmin(self.pixel_impedance, axis=0)
-
-    @property
-    def pixel_impedance_individual_offset(self) -> np.ndarray:
-        """Return the pixel impedance with the baseline of each individual pixel removed.
-
-        Each pixel in the resulting array has a minimum value of 0.
-        """
-        return self.pixel_impedance - self.pixel_baseline
```
We decided a while ago to remove these. The naming is not very intuitive. Also, we wanted to have these values saved as ContinuousData rather than re-calculate them each time they are needed.

psomhorst commented on eitprocessing/datahandling/eitdata.py:
```diff
@@ -36,25 +33,27 @@ class is meant to hold data from (part of) a singular continuous measurement.
+        if isinstance(self.path, str):
+            self.path = Path(self.path)
+        elif isinstance(self.path, list):
+            self.path = [Path(p) for p in self.path]
```
Thanks for the suggestion. I refactored ensure_path_list into parse_path with the option to always return a list, or have single arguments be returned as is.

psomhorst commented on eitprocessing/datahandling/eitdata.py:
```diff
@@ -36,25 +33,27 @@ class is meant to hold data from (part of) a singular continuous measurement.
 
     @staticmethod
-    def ensure_path_list(path: str | Path | list[str | Path]) -> list[Path]:
+    def ensure_path_list(path: str | Path | list[str | Path] | list[str] | list[Path]) -> list[Path]:
```
I made this change because type checkers did not understand. I now fixed this by creating two overloads in b87f829d28216ea199025fc25e91c093ff093fad.

DaniBodor replied:
why not always output a list of paths, even if it has a length of 1? Is there any implementation where this would be problematic? It would definitely make the code more readable to avoid the overload methods here.

psomhorst commented on eitprocessing/datahandling/intervaldata.py:
```diff
@@ -0,0 +1,185 @@
+T = TypeVar("T", bound="IntervalData")
+
+
+class TimeRange(NamedTuple):
+    """A tuple containing the start time and end time of a time range."""
+
+    start_time: float
+    end_time: float
```
NamedTuples are like dataclasses, but even lighter. The reason to use them over a normal tuple is that you can address the items using either the name or a integer getter:

```python
tr = TimeRange(10, 20)
tr2 = TimeRange(end_time=20, start_time=10)

assert tr == tr2
assert tr[0] is tr.start_time
```

Good call on nomenclature. I'll rename to Interval.

psomhorst replied:
Apparently it is not possible to subclass a slice object. But we can work around this by adapting `__getitem__` for all time-sliceable objects. E.g.:

```python
class Something:
    def __getitem__(self, key):
        if isinstance(key, TimeRange):
            # use TimeRange as slice
```

psomhorst replied:
I created #203 to handle this.

DaniBodor replied:
Have you given this any thought?

> I do think we could give it some additional features though ((it may need to become a dataclass at that point, not sure how far the functionality of named tuple goes).
> I am thinking e.g. of some checks like that `start_time` is smaller than `end_time`, or no negative numbers are given (I am correct in thinking that that shouldn't be possible right?).
> Even then, I'm not positive what the added benefit of a separate class is over an internal checking function.

psomhorst replied:
It makes sense to add some initialisation checks, which would more easily be implemented in a dataclass using `__post_init__`. 

It makes sense to do these checks at initialisation to remove that burden from any class using TimeRanges.

I created #207.

psomhorst commented on eitprocessing/datahandling/intervaldata.py:
```diff
@@ -0,0 +1,185 @@
+
+    def select_by_time(  # noqa: C901
+        self,
+        start_time: float | None = None,
+        end_time: float | None = None,
```
The function now has the same arguments as the function in slicing.py. I would like to keep it that way. We could change them all, but I don't know whether that improves usability.

psomhorst commented on eitprocessing/datahandling/intervaldata.py:
```diff
@@ -0,0 +1,185 @@
+    ) -> Self:
+        """Return only period data that overlaps (partly) with start and end time.
+
+        Other types of data (e.g. ContinuousData and SparseData) support the start_inclusive and end_inclusive
+        arguments. PeriodData does not. That means that selection by time of PeriodData probably works slightly
+        different than other types of selecting/slicing data.
```
I removed the reference to other objects. I think the rest of the docstring should be sufficient to explain how selection by time works.

DaniBodor replied:
It still reads a bit awkward.

This is what chatgpt suggested as an alternative docstring, which I quite like:
```python
    """
    Create a new copy of the object, selecting data between start_time and end_time.

    This function returns a shortened copy of the object, containing data from the specified 
    start_time to end_time. 

    If `partial_inclusion` is set to `True`, any intervals that overlap with the start_time 
    or end_time are included in the selection, and their times are adjusted to fit within 
    the specified range. If `partial_inclusion` is `False`, intervals that overlap the start 
    or end times are excluded from the selection.

    For example:
    - Set `partial_inclusion` to `True` for cases like "set_driving_pressure" where you want 
      to include settings that were active before the start_time.
    - Set `partial_inclusion` to `False` for cases like "detected_breaths" where you want to 
      exclude partial data that doesn't fully fit within the time range.

    Note that the end_time is always included in the selection if it is present in the original 
    object.

    Args:
        start_time: The earliest time point to include in the copy.
        end_time: The latest time point to include in the copy.
        partial_inclusion: Whether to include intervals that overlap with the start_time or end_time.
        newlabel: A new label for the copied object.
    """
```

psomhorst commented on eitprocessing/datahandling/intervaldata.py:
```diff
@@ -0,0 +1,185 @@
+            partial_inclusion = self.partial_inclusion
+        newlabel = newlabel or self.label
+
+        if start_time is None:
+            start_time = self.time_ranges[0].start_time
+        if end_time is None:
+            end_time = self.time_ranges[-1].end_time
```
I think it would not be beneficial to have the meaning of an argument change with the number of arguments. You can use `data.t[:end_time]` if you want a quick way to only give the end time.

psomhorst commented on eitprocessing/datahandling/intervaldata.py:
```diff
@@ -0,0 +1,185 @@
+            return TimeRange(start_time_, end_time_)
+
+        has_values = self.values is not None
+        iter_values = self.values or itertools.repeat(None)
+
+        time_range_value_pairs: Any = zip(self.time_ranges, iter_values, strict=True)
```
The type later changes to filter and list. The type checker assumes it is not allowed to change from zip to filter and list. That's what Any is for. 

Alternatively we could:
- make new variables each time we apply a filter (less readable)
- convert everything to lists (probably fine, maybe a bit slower)

psomhorst commented on eitprocessing/datahandling/intervaldata.py:
```diff
@@ -0,0 +1,185 @@
+        if end_time is None:
+            end_time = self.time_ranges[-1].end_time
+
+        def keep_starting_on_or_before_end(item: tuple[TimeRange, Any]) -> bool:
+            time_range, _ = item
+            return time_range.start_time <= end_time
+
+        def keep_ending_on_or_after_start(item: tuple[TimeRange, Any]) -> bool:
+            time_range, _ = item
+            return time_range.end_time >= start_time
+
+        def keep_fully_overlapping(item: tuple[TimeRange, Any]) -> bool:
+            time_range, _ = item
+            if time_range.start_time < start_time:
+                return False
+            if time_range.end_time > end_time:
+                return False
+            return True
+
+        def replace_start_end_time(time_range: TimeRange) -> TimeRange:
+            start_time_ = max(time_range.start_time, start_time)
+            end_time_ = min(time_range.end_time, end_time)
+            return TimeRange(start_time_, end_time_)
```
I refactored this part.

The keep_starting_on_.... functions are now reduced to two static functions (keep_overlapping and keep_fully_overlapping). replace_start_end_time is now also a static function.

The code has become more compact, but a bit harder to understand. It requires `partial`, because there is no other way (except for lambda's) to pass arguments to functions when using `filter` and `map`.

psomhorst replied:
I think zipping and unzipping is the best way to go here. Alternatively, we would need to filter and then keep score of which indices are kept, and then select the values with those indices.

DaniBodor replied:
Thanks for looking into this. I actually don't think that compactness is more important than readability. I also made an attempt to [refactor ](https://github.com/EIT-ALIVE/eitprocessing/pull/168/files#r1625735887) it in a way that I think makes it more readable (and even more compact). See what you think.

psomhorst replied:
I reverted the commit, because it led to test failure.

I agree that compactness isn't always best. However, I really don't like `for` loops for filtering by appending accepted items to a new list. I would use either `filter()` or a list comprehension. 
I also don't like keeping a separate list of indices to then use that for selecting from a second list. 

I re-refactored the function using most of your ideas in d01912bf994db9aaf6f596bb9e5ee0d16d9903a0. What do you think?

DaniBodor replied:
If it works then it's fine, and maybe not worth the time messing around with more.

That being said, to me, this solution still looks more complicated than I think it needs to be. 
- You are still creating a separate list of indices to select the values. In fact, you are creating 3 temporary lists (numbered_filtered_intervals, indices, filtered_intervals) versus 1 in my solution (keep_indices).
- The `if` statement in 144 doesn't really serve a purpose, apart from that zip fails on empty iterators. If numbered_filtered_intervals were a plain list, you could do away with the if and then an empty list would return empty lists as well. 
  - note: `if len(x)` is the same as `if x` (if x is not an iterator, `if len(x)` returns an error). Normally this triggers a ruff warning, not sure why that didn't happen here.
  - similarly, instead of `if self.values is not None`, we can use the property that we made for this `if not self.has_values` (also in concatenate)
- I agree that list comprehension/filter is normally cleaner and definitely more efficient than a "classical" for loop and appending items. But in this case, at least to me, the fact that multiple steps need to be taken (using indices on a separate list, trimming) makes the for loop/appending easier to read.
  - Actually, I think a list comprehension is only marginally more efficient than the for loop. Your first solution, using a filter, is definitely best in terms of efficiency, just makes reading it difficult.
  - if you really don't want to append, then maybe do everything in a single list comprehension to avoid having to avoid the if/else part below? Or is that even more confusing?:
```python
        iter_values = self.has_values or itertools.repeat(None)
        filtered_interval_value_pairs = [
            (iter_values[i], self._replace_start_end_time(interval, selection_start, selection_end))
            for i, interval in enumerate(self.intervals)
            if self._keep_overlapping(interval, selection_start, selection_end, partial_inclusion)
        ]
        # the return should have `values = values if self.has_values else None` or something similar
```
- I now see that the reason my suggestion failed is that I had passed `partial_inclusion` to `_replace_start_end_time` which it doesn't expect.

psomhorst commented on eitprocessing/datahandling/intervaldata.py:
```diff
@@ -0,0 +1,185 @@
+                return False
+            return True
+
+        def replace_start_end_time(time_range: TimeRange) -> TimeRange:
+            start_time_ = max(time_range.start_time, start_time)
+            end_time_ = min(time_range.end_time, end_time)
+            return TimeRange(start_time_, end_time_)
```
Yes, I had thought about it. I don't think it makes sense to keep it. If we find a use case where it does, we can adapt.

psomhorst commented on tests/test_loading.py:
```diff
@@ -95,6 +95,9 @@ def test_load_partial(
     assert len(draeger2_part1) == cutoff
     assert len(draeger2_part2) == len(draeger2) - cutoff
     assert draeger2_part1 == draeger2[:cutoff]
+    pytest.skip(
+        "Tests below rely on proper functioning of select_by_time, which should be refactored before fixing these tests",
+    )
```
select_by_time was not fixed in #189, right?

DaniBodor replied:
yes. I indeed checked that as well after writing my comment, but forgot to delete it.

psomhorst commented on eitprocessing/datahandling/eitdata.py:
```diff
@@ -12,9 +12,6 @@
-if TYPE_CHECKING:
-    from eitprocessing.datahandling.eitdata import Vendor
-
 T = TypeVar("T", bound="EITData")
 
 
```
Good point! I'll make it 'impedance data' instead. The data itself are not images, but can be converted to images in many different ways.

psomhorst commented on eitprocessing/datahandling/intervaldata.py:
```diff
@@ -0,0 +1,185 @@
+
+
+@dataclass(eq=False)
+class IntervalData(Equivalence, HasTimeIndexer):
```
The reason I did not include this was that you can't slice with the same time axis. However, it does make sense to be able to, say, select the first 10 occurences of a certain interval. I added SelectByIndex and _sliced_copy().

psomhorst commented on eitprocessing/datahandling/sequence.py:
```diff
@@ -56,7 +58,7 @@ def time(self) -> np.ndarray:
             return self.eit_data["raw"].time
         if len(self.continuous_data):
-            return next(self.continuous_data.values())
+            return next(iter(self.continuous_data.values()))
```
next() is required because the `values()` view is not indexable (because technically dicts don't have an order). By wrapping it in an iter() you can grab the first item with next(). It does roughly the same as `list(self.continuous_data.values())[0]` without creating the entire list.

psomhorst replied:
This method never returns None. It either returns an array or raises an exception.

psomhorst commented on eitprocessing/datahandling/loading/draeger.py:
```diff
@@ -65,8 +65,8 @@ def load_from_single_path(
     time = np.zeros((n_frames,))
-    events = []
-    phases = []
+    events: list = []
+    phases: list = []
```
mypy expects type hints; it can deduce list, but not what is in the list. I expanded it with what is in the list, because I forget every time.

psomhorst commented on eitprocessing/datahandling/datacollection.py:
```diff
@@ -5,14 +5,16 @@
     from typing_extensions import Self
 
-V = TypeVar("V", EITData, ContinuousData, SparseData)
+V_classes = (EITData, ContinuousData, SparseData, IntervalData)
```
The classes themselves should be hardcoded if we want to use it as a TypeVar. I've changed it to `V = TypeVar("V", *V_classes)`, but that breaks the types.

I tried solving this by creating a protocol HasLabel, which only has an attribute label. However, this does not work with the check `not any(issubclass(data_type, cls) for cls in V_classes)`.

The obvious (?) solution to this problem is create a ABC, for instance "DataContainer", that is a superclass of EITData, SparseData, IntervalData, ContinuousData and any container we'd like to add later. We might be able to move some common methods there. DataContainer can then be used for type hinting here, without hard coding the actual What do you think?

psomhorst commented on eitprocessing/datahandling/loading/draeger.py:
```diff
@@ -85,36 +85,86 @@ def load_from_single_path(
                 events,
                 phases,
                 previous_marker,
-                first_frame,
```
Yes. Events and phases (which are now removed from EITData itself) used to have the frame number (index) as an attribute. Since `index` in this function is the frame number _since the first frame_, the actual frame number in the event or phase had to be `index + first_frame`.

psomhorst commented on eitprocessing/datahandling/loading/draeger.py:
```diff
@@ -85,36 +85,86 @@ def load_from_single_path(
-            events=events,
-            label="raw",
-            pixel_impedance=pixel_impedance,
-        ),
+    eit_data = EITData(
+        vendor=Vendor.DRAEGER,
+        path=path,
+        framerate=framerate,
+        nframes=n_frames,
+        time=time,
+        label="raw",
+        pixel_impedance=pixel_impedance,
     )
+    eit_data_collection.add(eit_data)
```
I guess. I don't really like the fact the label 'raw' has to be here twice, but it's fine both ways.

DaniBodor replied:
I created a commit (0c78723: style: collection nomenclature) for renaming <XXX>_data_collection to <XXX>_collection throughout the loading functions. Feel free to drop the commit if you disagree with this.

psomhorst replied:
I read your comment after writing a comment below. I'll drop this one, except for renaming `load_func`.

psomhorst commented on eitprocessing/datahandling/loading/draeger.py:
```diff
@@ -85,36 +85,86 @@ def load_from_single_path(
+        SparseData(
+            label="minvalues_(draeger)",
+            name="Minimum values detected by Draeger device.",
+            unit=None,
```
For now, these only mark the locations of minvalues. The actual values are not stored here.

We might to rethink things that are automatically calculated at load time. We might want to re-calculate them after concatenation, for instance. I created #204.

psomhorst commented on eitprocessing/datahandling/loading/draeger.py:
```diff
@@ -85,36 +85,86 @@ def load_from_single_path(
+        SparseData(
+            label="maxvalues_(draeger)",
+            name="Maximum values detected by Draeger device.",
+            unit=None,
```
See above.

psomhorst commented on eitprocessing/datahandling/loading/sentec.py:
```diff
@@ -31,14 +31,14 @@ def load_from_single_path(  # noqa: C901
         version = reader.uint8()
 
-        time = []
+        time: list[float] = []
```
mypy likes this 🤷‍♂️

For some reason, mypy was automatically turned on for me a while ago. I noticed it detects quite some things that ruff does not cover. Maybe we should discuss whether to use it, and which things to (not) check for.

psomhorst commented on eitprocessing/datahandling/loading/sentec.py:
```diff
@@ -122,11 +122,15 @@ def load_from_single_path(  # noqa: C901
     )
 
-    return eit_data_collection, DataCollection(ContinuousData), DataCollection(SparseData)
+    return {
+        "eitdata_collection": eit_data_collection,
+        "continuousdata_collection": DataCollection(ContinuousData),
+        "sparsedata_collection": DataCollection(SparseData),
+    }
```
I thought I had fixed this, but I think that was only in a quick fix in another branch.

psomhorst replied:
I added it, and also added #205 to remove this requirement.

psomhorst commented on eitprocessing/datahandling/loading/timpel.py:
```diff
@@ -129,31 +153,127 @@ def load_from_single_path(
+        SparseData(
+            label="minvalues_(timpel)",
+            name="Minimum values detected by Timpel device.",
+            unit=None,
```
See comment before

psomhorst commented on eitprocessing/datahandling/loading/timpel.py:
```diff
@@ -129,31 +153,127 @@ def load_from_single_path(
+    )
 
-    phases.sort(key=lambda x: x.index)
+    gi = continuousdata_collection["global_impedance_(raw)"].values
```
`values` is an attribute, not a method.

DaniBodor replied:
sorry, I was confused with the `.values()` method of dictionaries.

psomhorst commented on eitprocessing/datahandling/mixins/equality.py:
```diff
@@ -14,22 +14,43 @@ class Equivalence:
 
     # inspired by: https://stackoverflow.com/a/51743960/5170442
-    def __eq__(self, other: Self) -> bool:
+    def __eq__(self, other: object) -> bool:
```
Apparently this is the expected type hint for __eq__.

psomhorst commented on eitprocessing/datahandling/mixins/equality.py:
```diff
@@ -14,22 +14,43 @@ class Equivalence:
 
+    def _eq_dataclass(self, other: object) -> bool:
+        """Compare two dataclasses for equality."""
+        if not is_dataclass(self) or not is_dataclass(other):
+            msg = "self or other is not a Dataclass"
+            raise TypeError(msg)
```
This is another one from mypy. These lines narrow the expected types that are used as input for `fields()` later. 

I agree that these seam superfluous.

DaniBodor replied:
ok, let's just leave it then. Not much overhead added here and extra insurance doesn't hurt in case things get refactored later.

psomhorst commented on eitprocessing/datahandling/mixins/slicing.py:
```diff
@@ -54,12 +55,15 @@ def select_by_index(
-        start = start or 0
-        end = end or len(self)
-        newlabel = newlabel or f"Slice ({start}-{end}] of <{self.label}>"
+        start = start if isinstance(start, int) else 0
+        end = end if isinstance(end, int) else len(self)
```
In the previous iteration, if `end = 0`, it was set to `len(self)` because `bool(0) is False`. (This was one I had to dig deep for while fixing the test.) `end` is 0 in some cases, which should result in an empty copy.

For `start` it does not really matter, but I added it for consistency.

DaniBodor replied:
I understand and good catch! I can imagine that that took a while to find 😖.

Is this marginally better? It protects against a user inputting something random that is not supposed to be passed here. (Note: for start it also protects against something falsy like an empty string, which `start or 0` would not).

```suggestion
        start = start if start is not None else 0
        end = end if end is not None else len(self)
```

psomhorst commented on eitprocessing/datahandling/mixins/slicing.py:
```diff
@@ -128,10 +135,14 @@ def select_by_time(  # noqa: D417
             raise TypeError(msg)
 
+        if len(self.time) == 0:
+            # TODO: make proper new instances when not slicing
```
Maybe?

However, there are certainly times I would not want to raise an error when a sliced object becomes empty. For instance: 

We have a Sequence with IntervalData containing breaths holds at t=12-15 and t=22-25. If we want to split this into multiple sub-sequences from 0-9, 10-19 and 20-29 seconds, the first object does not get any of the breath holds. This should not raise an exception. Also, I think the IntervalData object itself should stay intact and just be empty (intervals will become an empty list), such that you can apply the same logic to each sub-sequence.

psomhorst commented on eitprocessing/datahandling/loading/timpel.py:
```diff
@@ -129,31 +153,127 @@ def load_from_single_path(
+    time: np.ndarray,
+    min_indices: np.ndarray,
+    max_indices: np.ndarray,
+    gi: np.ndarray,
```
gi is the `values` attribute from a ContinuousData object. This is always an np.ndarray (and is typed as such). 
The typo above was not a typo.

DaniBodor replied:
my bad!

psomhorst commented on eitprocessing/datahandling/loading/timpel.py:
```diff
@@ -129,31 +153,127 @@ def load_from_single_path(
+    }
+
+
+def _make_breaths(
```
This is absolutely correct. As soon as this PR is done, BreathDetection is one of my next projects. This is almost finished, and also contains this code. I'll get to that when finishing BreathDetection.

DaniBodor commented on eitprocessing/datahandling/intervaldata.py:
```diff
@@ -0,0 +1,241 @@
+            values=list(values) if self.has_values else None,
+        )
+
+    @staticmethod
+    def _keep_overlapping(
+        item: tuple[Interval, Any],
+        selection_start: float,
+        selection_end: float,
+    ) -> bool:
+        """Helper function for filtering overlapping interval-value pairs."""
+        interval, _ = item
+        return interval.start_time <= selection_end and interval.end_time >= selection_start
+
+    @staticmethod
+    def _keep_fully_overlapping(
+        item: tuple[Interval, Any],
+        selection_start: float,
+        selection_end: float,
+    ) -> bool:
+        """Helper function for filtering fully overlapping interval-value pairs."""
+        interval, _ = item
+        if interval.start_time < selection_start:
+            return False
+        if interval.end_time > selection_end:
+            return False
+        return True
```
```suggestion
    @staticmethod
    def _keep_overlapping(
        item: tuple[Interval, Any],
        selection_start: float,
        selection_end: float,
        keep_partial_overlapping: bool = True,  # not sure whether or in which direction this should be defaulted.
    ) -> bool:
        """Helper function for filtering overlapping interval-value pairs."""
        interval, _ = item
        if keep_partial_overlapping:
            return interval.start_time <= selection_end and interval.end_time >= selection_start
        if interval.start_time < selection_start:
            return False
        if interval.end_time > selection_end:
            return False
        return True
```

DaniBodor commented on eitprocessing/datahandling/intervaldata.py:
```diff
@@ -0,0 +1,236 @@
+
+        filterfunc = partial(
+            self._keep_overlapping,
+            selection_start=selection_start,
+            selection_end=selection_end,
+            keep_partial_overlapping=partial_inclusion,
+        )
+        filtered_pairs = list(filter(filterfunc, interval_value_pairs))
+
+        if len(filtered_pairs):
+            intervals, values = zip(*filtered_pairs, strict=True)
+            mapfun = partial(self._replace_start_end_time, selection_start=selection_start, selection_end=selection_end)
+            intervals = list(map(mapfun, intervals))
+        else:
+            intervals, values = [], []
+
+        return type(self)(
+            label=newlabel,
+            name=self.name,
+            unit=self.unit,
+            category=self.category,
+            derived_from=[*self.derived_from, self],
+            intervals=list(intervals),
+            values=list(values) if self.has_values else None,
+        )
+
+    @staticmethod
+    def _keep_overlapping(
+        item: tuple[Interval, Any],
+        selection_start: float,
+        selection_end: float,
+        keep_partial_overlapping: bool,
+    ) -> bool:
+        """Helper function for filtering overlapping interval-value pairs."""
+        interval, _ = item
+
+        if keep_partial_overlapping:
+            return interval.start_time < selection_end and interval.end_time > selection_start
+
+        return interval.start_time >= selection_start and interval.end_time <= selection_end
```
I think this does the same thing without the need for partials, etc. I think this is much easier to parse for humans.
It does contain a loop and a list comprehension rather than the filter you made, so probably less efficient. But I think that improved readability compensates for what is probably not a bottleneck in speed/efficiency.
Note that code below is untested.

Also, can we change nomenclature of all `select_by_time` functions to `selection_start` and `selection_end` rather than doing it on the spot just for this one. I think in general, it creates less confusion to have a given keyword have a single function throughout the codebase. Now `start_time ` can either be part of a `select_by_time` function or of an `Interval`. (Alernatively, we can make it `Interval.interval_start` and `Interval.interval_end`, but that looks a bit silly).

```suggestion
        selection_start = start_time or self.intervals[0].start_time
        selection_end = end_time or self.intervals[-1].end_time
        partial_inclusion = partial_inclusion or self.default_partial_inclusion
        newlabel = newlabel or self.label

        # return copy of self if neither start nor end is selected
        if start_time is None and end_time is None:
            copy_ = copy.deepcopy(self)
            copy_.derived_from.append(self)
            copy_.label = newlabel
            return copy_

        selection_conditions = (selection_start, selection_end, partial_inclusion)
        keep_intervals, keep_indices = [], []
        for i, interval in enumerate(self.intervals):
            if self._keep_overlapping(interval, *selection_conditions):
                keep_intervals.append(self._replace_start_end_time(interval, *selection_conditions))
                keep_indices.append(i)
        keep_values = [self.values[x] for x in keep_indices] if self.has_values else None

        return type(self)(
            label=newlabel,
            name=self.name,
            unit=self.unit,
            category=self.category,
            derived_from=[*self.derived_from, self],
            intervals=keep_intervals,
            values=keep_values,
        )

    @staticmethod
    def _keep_overlapping(
        interval: Interval,
        selection_start: float,
        selection_end: float,
        partial_inclusion: bool,
    ) -> bool:
        """Helper function for filtering overlapping interval-value pairs."""
        if partial_inclusion:
            return interval.start_time <= selection_end and interval.end_time >= selection_start
        return interval.start_time >= selection_start and interval.end_time <= selection_end
```

psomhorst replied:
I addressed this in some other comment.

DaniBodor commented on eitprocessing/datahandling/sparsedata.py:
```diff
@@ -13,38 +13,40 @@
-    In sparse data the time points are not necessarily evenly spaced. Data can consist time-value pairs or only time
-    points. Values generally are numeric values in arrays, but can also be lists of different types of object.
+    Sparse data is data for which the time points are not necessarily evenly spaced. Data can consist time-value pairs
+    or only time points. Values generally are numeric values in arrays, but can also be lists of different types of
+    object.
```
I don't think the comment on values adds clarity here, given that I don't think there are any true restrictions (neither programmatically nor in practice).

```suggestion
    or only time points.
```

psomhorst replied:
I think I added this here to contrast ContinuousData where `values` is expected to be a numpy array. But it is fine not to mention it specifically.

DaniBodor commented on eitprocessing/datahandling/sparsedata.py:
```diff
@@ -13,38 +13,40 @@
+    object.
 
-    Sparse data differs from IntervalData in that each data points is associated with a single time point rather than a
+    Sparse data is assumed to be sequential (i.e. only a single data points at a single time point, and sorted by time).
+    Sparse data can contain zero or more time points/time-values pairs.
```
A few comments:
- I think it would be good to either enforce that it is sequential or not comment on it (idem for interval data).
- The part between brackets is unclear to me.
- I think the last sentence can be omitted (if not, it should be included for the other containers as well).

psomhorst replied:
I removed both sentences. 

We currently have almost no checks for values that are set. This might be useful to do at some point, but not part of this PR. 

I previously added the second sentence because you commented you weren't sure whether SparseData was a single or more than one time point. It can be any number of time points.

DaniBodor replied:
> I previously added the second sentence because you commented you weren't sure whether SparseData was a single or more than one time point. It can be any number of time points.

I understand. I think that's my bad. I think it's more a generic confusion I keep reverting to and then remembering regarding data collections and data containers and data sets rather than about sparse data per se. It would be good to at some point create clear documentation defining what each of these is (for human comprehension).

DaniBodor replied:
> We currently have almost no checks for values that are set. This might be useful to do at some point, but not part of this PR.

Agreed. I've added a comment to [#208 comment](https://github.com/EIT-ALIVE/eitprocessing/issues/208#issuecomment-2149397241)

DaniBodor reviewed (changes requested):
Thanks for the changes, I think it's improved quite a bit.
I still added a few suggestions, and will also push a few commits I made for cases where that is clearer than a "code suggestion" in GitHub. Nonetheless, these commits are still intended as suggestions, so feel free to drop or change them as you please.

psomhorst commented:
I reverted 0c7872343747defa58e1ffdec233fad8fe040a7e. It introduced bugs resulting in failed tests.

psomhorst commented:
@DaniBodor I don't really understand b208e09f731de810b3723bbb06b63fa7fb7af354. "sparse_collection" is less clear to me than "sparsedata_collection". Reading the code, it is not very clear what the relationship between a `SparseData` object and the name "sparse_collection" is. Of course the same goes for the other data types.

I'm not bothered by the slightly longer names. Shortening names does not save any lines, only characters.

psomhorst commented:
@DaniBodor I addressed most/all (?) of your comments. GitHub is making quite a mess of them when they get outdated, so sorry if I missed one. 

The most important one is the refactor of `IntervalData.select_by_time()`. I didn't really like your solution, but took inspiration from it and refactored it again to remove `filter()`, `map()` and `partial()`. It now uses list comprehensions instead.

DaniBodor commented:
> @DaniBodor I addressed most/all (?) of your comments. GitHub is making quite a mess of them when they get outdated, so sorry if I missed one.
> 
> The most important one is the refactor of `IntervalData.select_by_time()`. I didn't really like your solution, but took inspiration from it and refactored it again to remove `filter()`, `map()` and `partial()`. It now uses list comprehensions instead.

Thanks, I think it's fine too. Also lost track already. I looked at the new commits and think all is fine like this (see my comment on the refactoring back/and forth). If we do want to change that, it can also become a new PR if needed.

Can you please fix the minor linting errors (you may need to update ruff first, then `ruff check` and `ruff format`).

Did you still want to add tests to this PR, or will that be a separate one?

psomhorst commented:
@DaniBodor 
- I addressed the linting errors
- I added tests for IntervalData and SparseData, with 100% coverage and 100% success rate
- Only `SparseData.select_by_time()` will have to be tested after we fix slicing by time. There are tests with the proper names, but they are skipped.

I think this is now done. If you agree, we can merge it.

psomhorst commented:
@DaniBodor I think this is now finally done. The linting fails on GitHub, but I don't see that same linting error on my computer. Is that a local configuration error, or is something else going on?

DaniBodor commented on tests/test_intervaldata.py:
```diff
@@ -0,0 +1,196 @@
+    assert not intervaldata1.has_values
+    intervaldata1.values = []
+    assert intervaldata1.has_values
+    intervaldata1.values = None
```
just double checking that the reverse operation also updates the property.
```suggestion
    intervaldata1.values = None
    assert not intervaldata1.has_values
```

DaniBodor commented on tests/test_intervaldata.py:
```diff
@@ -0,0 +1,196 @@
+
+
+def test_index_slicing(intervaldata1: IntervalData) -> None:
+    _sliced_copy = intervaldata1._sliced_copy(0, 10, newlabel="sliced_copy")  # noqa: SLF001
```
why did we make `_sliced_copy` a private method? I think it would be nice to have as a public method, in case users want to add a custom label.
If we want to change this, I suggest to do it in a separate PR.

psomhorst replied:
Good question. I see no reason to keep it private. I created #212.

DaniBodor commented on tests/test_intervaldata.py:
```diff
@@ -0,0 +1,196 @@
+
+def test_select_by_time(intervaldata1: IntervalData, intervaldata2: IntervalData) -> None:
+    assert intervaldata1.t[:1] == intervaldata1[:1]
+
```
here is where it would be nice to see the functional difference from the variable name :)
Then it's immediately clear why the expectations differ between the two.

DaniBodor commented on tests/test_intervaldata.py:
```diff
@@ -0,0 +1,196 @@
+def test_select_by_time(intervaldata1: IntervalData, intervaldata2: IntervalData) -> None:
+    assert intervaldata1.t[:1] == intervaldata1[:1]
+
+    assert len(intervaldata1.t[:1]) == 1
+    assert len(intervaldata1.t[:0.5]) == 1
+
+    assert len(intervaldata2.t[:1]) == 1
+    assert len(intervaldata2.t[:0.5]) == 0
```
for these it would be nice to juxtapose the different ones, as was done below.

DaniBodor commented on tests/test_intervaldata.py:
```diff
@@ -0,0 +1,196 @@
+    assert len(sliced_copy) == 3
+
+    sliced_copy = intervaldata1.t[:]
+    assert len(sliced_copy) == len(intervaldata1)
```
they should be considered equal, not just in length

```suggestion
    assertsliced_copy == intervaldata1
```

DaniBodor commented on tests/test_intervaldata.py:
```diff
@@ -0,0 +1,196 @@
+@pytest.fixture()
+def intervaldata1():
+    """Creates an IntervalData object with n sequential intervals lasting 1 second each."""
+    n = random.randint(50, 150)
```
why not just fix the `n` in these tests?
Same question for the values in the fixtures below.

psomhorst replied:
You are probably correct that it does not matter in this case. I have had some situations in the past where the tests worked fine because of the specific length of the object I was testing. However, there are probably very little side effects with a fixed length.

I made them all different lengths, though, to prevent accidents.

DaniBodor commented on tests/test_sparsedata.py:
See my comments on test_intervaldata.
In addition:
- the docstring of the fixtures still refers to intervals
- sparsedata2 fixture is identical to sparsedata1, so can be removed.

DaniBodor commented on eitprocessing/datahandling/sparsedata.py:
```diff
@@ -1,2 +1,122 @@
+        end_index: int,
+        newlabel: str,
+    ) -> Self:
+        # TODO: check correct implementation
```
done now, i believe :)
```suggestion
```

DaniBodor reviewed (changes requested):
Thanks for adding the tests!
I made some minor suggestions to them, main one is about functionally naming the fixtures

I think this will be the last round of changes and then done :)

psomhorst commented on eitprocessing/datahandling/sparsedata.py:
```diff
@@ -1,2 +1,122 @@
+        cls = type(self)
+        newlabel = newlabel or self.label
+
+        new_values: Any
+        if isinstance(self.values, list) and isinstance(other.values, list):
+            new_values = self.values + other.values
+        elif isinstance(self.values, np.ndarray) and isinstance(other.values, np.ndarray):
+            new_values = np.concatenate((self.values, other.values))
+        elif self.values is None and other.values is None:
+            new_values = None
+        else:
+            msg = "Value types of self and other do not match."
+            raise TypeError(msg)
```
Great suggestion! Thanks!

psomhorst commented:
@DaniBodor Thanks for the review! 

Great suggestions on the concatenation type check. Also great suggestions for the readability of the tests. 

GitHub says there are still unresolved issues. However, I can't find them anywhere. Is there anything still outstanding?
